rancher/dashboard
# Add catalog list/loading for air-gapped extensions

### Summary

This adds a new option to load Extension images from a specific registry image and load multiple packages as unique charts.

#### Additions
- "Images" tab in the Extensions page
  - Contains ResourceList for loaded container images
- Dialog for loading image with secrets
- Github workflow for building  Extension container image

#### Loading process
The image loading process consists of 4 parts:
- Creating the deployment from the extension image
- Exposing the deployment with a ClusterIP service
- Creating a Helm repository from the service's ClusterIP
- Creating the UIPlugin crd for the Extension.

Once all of the resources have been successfully created, this will prompt a reload from the user. Then the charts listed under the created Helm repository will be shown in the "Available" tab.

#### Extension build process
The build process for the container image of an extension has been modified to allow the root Extension to act as a Helm repository and each package within the `./pkg` directory to be listed as charts. 

Also the dependency on the [ui-plugin-server](https://github.com/rancher/ui-plugin-server) has been removed and placed into the `./shell/scripts/extension/helm` directory. Instead of cloning the repository during the `publish` script, the contents of the `helm` directory is copied and iterated over for each package. This directory contains a condensed version of the chart originally used (`./helm/charts`), the Dockerfile and nginx.conf for the image build (`./helm/package`), and the scripts for used to populate the config for the build (`./helm/scripts`).

#### Extension workflow
A new workflow (`build-container.yml`) has been added to the `@rancher/pkg` creator. This will build the extension as a container image and push it into the Github Container Registry (`ghcr.io`). This will make the image easily accessible for a user to tag and push into a specified registry.  

### To test
Follow the documentation in `./docusaurus/docs/extensions/advanced/air-gapped-environments.md`. This [container image](https://github.com/jordojordo/extensions-test-pt-deux/pkgs/container/ui-extension-extensions-test-pt-deux/94081344?tag=0.1.2) can be used to test, it contains 2 top-level extension packages. If necessary, I can provide private registry as well.

1. Load image
![image-tab](https://github.com/rancher/dashboard/assets/40806497/4f376ca0-602f-491a-9160-4e4fdcebfba5)
Input image url
![image-load-dialog](https://github.com/rancher/dashboard/assets/40806497/17306996-6eba-4fe4-bfd8-dd29eaf4c91a)

2. Reload when prompted
![reload](https://github.com/rancher/dashboard/assets/40806497/aa218f16-d46f-452c-91eb-2a82d72f739c)

3. See cache state updated for image
![cached](https://github.com/rancher/dashboard/assets/40806497/4dad97ab-13a8-4035-ae2d-0add9913a75d)

4. Install one of the new extensions in the Available tab
![available](https://github.com/rancher/dashboard/assets/40806497/48f2cf01-234f-4285-885a-2e24506b5df7)

5. Once installed navigate to the new menu item within the side-nav
![new-page](https://github.com/rancher/dashboard/assets/40806497/0d12ba7d-6b7e-4656-8a9a-5f7269c1157a)
